Add better error messages in case of some parsing problems.

hiker opened the pull request:
Some minor modification so that a user gets better error message (esp. in psyclone - instead of just "Is the Fortran correct").

arporter commented on src/fparser/api.py:
```diff
@@ -67,19 +67,18 @@
+from .base_classes import classes
+
+from .utils import AnalyzeError
+
```
Tidying of imports - nice.

arporter commented on src/fparser/base_classes.py:
```diff
@@ -853,15 +853,16 @@ def process_subitem(self, item):
         self.content.append(item)
-        #sys.exit()
+        raise AnalyzeError(message)
+        # sys.exit()
```
I think we could delete this commented-out exit() call now.

arporter replied:
New code is covered but `return` statement below is now not required.

arporter commented on src/fparser/base_classes.py:
```diff
@@ -853,15 +853,16 @@ def process_subitem(self, item):
         return
 
     def handle_unknown_item(self, item):
```
Please document the interface to this routine.

arporter commented on src/fparser/parsefortran.py:
```diff
@@ -118,22 +118,27 @@ def put_item(self, item):
         try:
-            block = self.block = BeginSource(self)
+            self.block = BeginSource(self)
+        except AnalyzeError:
```
Catch the AnalyzeError that the parser can now raise - good.

arporter replied:
This new code is covered by the test suite - good.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -810,23 +810,30 @@ def format_message(self, kind, message, startlineno, endlineno,
+    def format_warning_message(self, message, startlineno=None,
+                               endlineno=None, startcolno=0, endcolno=-1):
+        '''Create a string with a warning message.'''
+        if startlineno is None:
```
Please add a comment to explain why/how you adjust start/endlineno.

arporter replied:
In fact, these lines aren't covered by the test suite either. Please can you add/modify tests such that they are.

arporter replied:
Thanks for adding the docstring. However, the new lines of code are not covered. Also, what happens if `startlineno=None` and `self.linecount < 2`?

hiker replied:
This confuses me a bit ... while I agree that in my original pull request I had not handled that case, but I've fixed that, see: https://github.com/stfc/fparser/pull/55/files:
 +      if startlineno is None:
+            startlineno = self.linecount-2
+            if startlineno < 0:
+                startlineno = 1

The change above is marked as 'outdated' - I assume there was a mis-click somewhere ;)

arporter commented on src/fparser/api.py:
```diff
@@ -216,13 +215,18 @@ def parse(input, isfree=None, isstrict=None, include_dirs = None, source_only =
+    if not parser.block:
+        message = reader.format_warning_message("Parser Error")
+        raise AnalyzeError(message)
+
```
Also, these lines are not covered by the test suite - please add a test for them.

arporter commented on src/fparser/parsefortran.py:
```diff
@@ -118,22 +118,27 @@ def put_item(self, item):
-                                                reader.linecount, reader.linecount)
+                message = \
+                    reader.format_message('FATAL ERROR',
+                                          'while processing line',
```
This code is not covered but as it has only been tidied for pep8 I won't request it.

hiker commented:
Hi,
thanks for the feedback. My pull request might need a bit of an explanation: no, I did not just remove the line that had no test case. It turns out that when this line would be triggered, the original parser message was lost. So instead I am now raising an exception, so that the error message can properly be handled.
One additional question: what is the best way to check the coverage? I can see see the link in this pull request after I push to my repo, but I am sure there's a command-line-way of checking that before I push.
Thanks!
   Joerg

rupertford commented:
For command line checking I use

` py.test --cov-report term-missing --cov fparser.xxx`
 where xxx is the python module name.

It requires an extension to py.test which I believe is installed thus ...

`sudo pip install pytest-cov`

arporter commented:
My preferred output is html as it makes cross-referencing with the changes (displayed in a github PR) much easier:

    py.test --cov-report html ...

arporter commented on src/fparser/readfortran.py:
```diff
@@ -797,14 +797,15 @@ def format_message(self, kind, message, startlineno, endlineno,
                 sourceline = self.source_lines[i-1]
                 l0 = linenostr+sourceline[:startcolno]
-                if endcolno==-1:
+                if endcolno == -1:
```
Tidying for pep8 - thanks!

arporter commented on src/fparser/tests/test_parser.py:
```diff
@@ -869,3 +870,14 @@ def test_invalid_type_bound_array_access():  # pylint: disable=invalid-name
     assert "with Call pattern failed" in str(excinfo)
+
+
+def test_analyze_errors():
```
New test - great!

hiker commented:
Sorry, wasn't entirely ready - since I needed to check coverage (which I didn't know how to do) - plus the link on the web interface didn't work for me this time (I get: "Source Not Available   The file "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/fparser/readfortran.py" isn't available on github. Either it's been removed, or the repo root directory needs to be updated.).

But thanks for the feedback, but will be a few days before I can get to that I guess.

arporter commented:
Oops - sorry for jumping the gun in that case. I shall wait till you tell me you're ready next time! That message about missing source is because the test suite now runs as two sub-jobs in Travis - one for Python 2.7 and one for Python 3.something. Things are currently set-up so that the source browser will find the source for the Python3.something tests but not the 2.7. I haven't investigated to see whether it's possible to get it to work for both.

hiker commented:
Don't worry about the link, since you gave me the info how to do it myself I don't need them anymore ;)

hiker commented:
OK, next try. I was quite confused by the 'new code not covered by test case' (format_warning or so function), since it was covered when I tested it. But turhs out that when I cleaned up some code based on your initial comments, I actually removed the code that would have triggered the format_warning call in the first place - so it was safe to just remove the code.
And while I usually avoid adding more new stuff in a pull-request fix, there were some really low hanging fruits and I added two or three easy tests.
The travis errors are apparently only python3 related, so I assume I can ignore that?

Hope that's all now :)

Thanks!
   Joerg

arporter commented:
Thanks @hiker, I'm afraid ignoring failing tests (even for Python 3) is not an option :-) Please could you fix and then I'll review...

hiker commented:
Oops - misunderstanding. I thought that python3 is not yet supported (and as part of an upcoming port someone had set up python3 test ... never realised that they were running all the time). OK, I tried to fix it (though not sure how to trigger the python3 tests locally yet).
